heron/api/src/java/com/twitter/heron/api/bolt/IStatefulBolt.java

```diff
@@ -0,0 +1,27 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.api.bolt;
+
+import com.twitter.heron.api.topology.IStatefulComponent;
+
+/**
+ * An IStatefulBolt represents a bolt that implements stateful functionality.
+ * In addition to the regular prepare/execute methods, it has initialization
+ * methods with state as defined in IStatefulComponent. An implementation
+ * of the stateful bolt saves state in the State interface and is able
+ * to restore its state from the State during init call.
+ */
+public interface IStatefulBolt extends IStatefulComponent, IRichBolt {
```

Comment:
I would think having IStatefulComponent is good enough; no need IStatefulBolt or IStatefulSpout:
- IStatefulBolt does not add more methods than IStatefulComponent and IRichBolt. A bolt can simply implement IStatefulComponent if it wants Stateful feature; on the contrast, IUpdatable interface looks better (https://github.com/twitter/heron/blob/master/heron/api/src/java/com/twitter/heron/api/topology/IUpdatable.java).
- IStatefulBolt extends IRichBolt. Then how about IBasicBolt? Need we add it too? And how about others, ? It can either lead to an explosion of APIs, which in fact IMO are duplicated; or it can lead to more refactoring on existing topologies to support these features (for instance, current bolt extends IBaseBasicBolt, which provides different APIs from IRichBolt).

In a word, I think this wrapper complicates things without bringing much benefits.